lib/compat/wordpress-6.5/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -673,6 +674,92 @@ private function data_wp_text_processor( WP_Interactivity_API_Directives_Process
 				}
 			}
 		}
+
+		/**
+		 * Processes the `data-wp-router-region` directive.
+		 *
+		 * It renders in the footer a set of HTML elements to notify users about
+		 * client-side navigations. More concretely, the elements added are 1) a
+		 * top loading bar to visually inform that a navigation is in progress
+		 * and 2) an `aria-live` region for accessible navigation announcements.
+		 *
+		 * The same action is registered everytime the directive is processed to
+		 * prevent element duplication.
+		 *
+		 * @since 6.5.0
+		 *
+		 * @param WP_Interactivity_API_Directives_Processor $p The directives processor instance.
+		 */
+		private function data_wp_router_region_processor( WP_Interactivity_API_Directives_Processor $p ) {
+			if ( ! $p->is_tag_closer() ) {
+				/*
+				 * The state could be declared multiple times here but is
+				 * doesn't matter as the values are always the same.
+				 */
+				wp_interactivity_state(
+					'core/router',
+					array(
+						'navigation' => array(
+							'message'     => '',
+							'hasStarted'  => false,
+							'hasFinished' => false,
+							'texts'       => array(
+								'loading' => __( 'Loading page, please wait.' ),
+								'loaded'  => __( 'Page Loaded.' ),
+							),
+						),
+					)
+				);
+
+				$callback = static function () {
+					echo <<<HTML
+<style>
+.wp-interactivity-router_loading-bar {
+	position: fixed;
+	top: 0;
+	left: 0;
+	margin: 0;
+	padding: 0;
+	width: 100vw;
+	max-width: 100vw !important;
+	height: 4px;
+	background-color: var(--wp--preset--color--primary, #000);
+	opacity: 0
+}
+.wp-interactivity-router_loading-bar.start-animation {
+	animation: wp-interactivity-router_loading-bar-start-animation 30s cubic-bezier(0.03, 0.5, 0, 1) forwards
+}
+.wp-interactivity-router_loading-bar.finish-animation {
+	animation: wp-interactivity-router_loading-bar-finish-animation 300ms ease-in
+}
+
+@keyframes wp-interactivity-router_loading-bar-start-animation {
+	0% { transform: scaleX(0); transform-origin: 0% 0%; opacity: 1 }
+	100% { transform: scaleX(1); transform-origin: 0% 0%; opacity: 1 }
+}
+@keyframes wp-interactivity-router_loading-bar-finish-animation {
+	0% { opacity: 1 }
+	50% { opacity: 1 }
+	100% { opacity: 0 }
+}
+</style>
+<div
+	class="wp-interactivity-router_loading-bar"
+	data-wp-interactive='{"namespace":"core/router"}'
+	data-wp-class--start-animation="state.navigation.hasStarted"
+	data-wp-class--finish-animation="state.navigation.hasFinished"
+></div>
+<div
+	class="screen-reader-text"
+	aria-live="polite"
+	data-wp-interactive='{"namespace":"core/router"}'
+	data-wp-text="state.navigation.message"
+></div>
+HTML;
+				};
+				add_action( 'wp_footer', $callback );
```

Comment:
Have you tested that this action adds the markup only once, even if it's called multiple times? I'm not sure what it does with references instead of strings.

Reply:
I used a static variable to know if the `wp-router-region` processor already set a hook in `wp_footer`. There are also tests to ensure the HTML is not duplicated when there are more than one `wp-router-region` directives.

Reply:
Well, I ended up using a private property inside the `WP_Interactivity_API` class (see https://github.com/WordPress/gutenberg/pull/58377/commits/bd3842c8290e53afd8edbb3172ee4748499f6a25).

The issue is that the static variable is shared between any `WP_Interactivity_API` instance regardless of whether they have processed a `wp-router-region` directive, making other instances not output the markup.

Reply:
Nice, thank you.